Refactor CommandProcessor-implementing Job processors

koevskinikola opened the pull request:
Refactor the `JobThrowErrorProcessor` and `JobCancelProcessor` classes to implement the `TypedRecordProcessor` instead of the `CommandProcessor` interface.

ℹ️ I'll backport this PR to avoid issues with future backports in these processors.

## Description

<!-- Describe the goal and purpose of this PR. -->

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [x] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes to #40924

koevskinikola committed:
refactor: job cancel processor implements typed record processor

Refactor the JobCancelProcessor to implement the
TypedRecordProcessor instead of the CommandProcessor.

koevskinikola committed:
refactor: job throw error processor implements typed record processor

Refactor the JobThrowErrorProcessor to implement the
TypedRecordProcessor instead of the CommandProcessor.

koevskinikola committed:
refactor: remove command processor reference in exception javadocs

koevskinikola commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobCancelProcessor.java:
```diff
@@ -19,32 +23,38 @@
     } else {
-      commandControl.reject(RejectionType.NOT_FOUND, String.format(NO_JOB_FOUND_MESSAGE, jobKey));
+      rejectionWriter.appendRejection(
+          record, RejectionType.NOT_FOUND, NO_JOB_FOUND_MESSAGE.formatted(jobKey));
```
@copilot jobs can't be cancelled by a user, so no resonse is necessary.

remcowesterhoud replied:
🙃 @koevskinikola Copilot has a point though, we do write a response if the job can be found. The response won't go anywhere, as there won't be any request data but it's odd that we have a differentiation in this processor. How about removing the other one?

koevskinikola committed:
refactor: job complete processor implements typed record processor

Remove usages of the CommandProcessor interface so that
it can be removed. The CommandProcessor interface is
is difficult to understand and is a frequent source of
processor-related bugs.

This commit also removes the DefaultJobPreconditionGuard
class and JobAccept interfaces as they aren't used anymore.

remcowesterhoud commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobThrowErrorProcessor.java:
```diff
@@ -152,30 +162,46 @@ private void acceptCommand(
+      return;
+    }
+
+    eventPublicationBehavior.throwErrorEvent(foundCatchEvent.get(), job.getVariablesBuffer());
```
❌ We shouldn't throw the error event in a method called `raiseIncident`. That's a very unexpected side-effect.

koevskinikola committed:
refactor: don't throw error event if command is rejected or no event found

If a command is rejected stop the execution of the
throwError method.

If no catch event is found, raise an incident, and
only if a catche event is found, invoke the event
publication behavior.

remcowesterhoud reviewed (approved):
Thanks for the changes!

koevskinikola committed:
refactor: ensure job metrics on throw error are tracked

koevskinikola commented:
Hey @mustafadagher, I'm aware, I need to refactor them to adjust for the changes between versions.
